File: typescript/packages/core/src/agent/cache-ttl.ts
import type { Model } from "../model";
import { ConfigError } from "./errors";

// The default context.cache_ttl_ms from the models' declared cache lifetimes (ModelInfo.cache).
// cacheBreaks() judges every response of a thread by one TTL, so the default must hold for every
// model the thread can use: the primary and its fallbacks.

/**
 * The TTL every caching model declares, or undefined when none caches. Throws invalid_config
 * when a model's lifetime is unknown or two models disagree: context.cache_ttl_ms must decide.
 */
export function agreedCacheTtl(models: readonly Model[]): number | undefined {
  let agreed: { readonly model: Model; readonly ttl: number } | undefined;
  for (const model of models) {
    const { cache } = model.info;
    if (cache === "none") continue;
    if (cache === undefined)
      throw new ConfigError(
        "invalid_config",
        `the cache lifetime of ${name(model)} is unknown: set context.cache_ttl_ms, or pass cacheTtlMs to its factory`,
      );
    if (agreed === undefined) agreed = { model, ttl: cache.ttl_ms };
    else if (agreed.ttl !== cache.ttl_ms)
      throw new ConfigError(
        "invalid_config",
        `${name(agreed.model)} caches for ${span(agreed.ttl)} but fallback ${name(model)} for ${span(cache.ttl_ms)}: set context.cache_ttl_ms`,
      );
  }
  return agreed?.ttl;
}

function name(model: Model): string {
  return `${model.info.model.provider}/${model.info.model.name}`;
}

function span(ms: number): string {
  if (ms % 3_600_000 === 0) return `${ms / 3_600_000}h`;
  if (ms % 60_000 === 0) return `${ms / 60_000}m`;
  return `${ms} ms`;
}

File: python/src/threads/agents/cache_ttl.py
"""The default context.cache_ttl_ms from the models' declared cache lifetimes (ModelInfo.cache).
cache_breaks() judges every response of a thread by one TTL, so the default must hold for every
model the thread can use: the primary and its fallbacks."""

from collections.abc import Iterable

from threads.agents.config import ConfigError
from threads.loop.model import Model


def agreed_cache_ttl(models: Iterable[Model]) -> int | None:
    """The TTL every caching model declares, or None when none caches. Raises invalid_config when
    a model's lifetime is unknown or two models disagree: context.cache_ttl_ms must decide."""
    agreed: tuple[Model, int] | None = None
    for model in models:
        cache = model.info.cache
        if cache == "none":
            continue
        if cache is None:
            raise ConfigError(
                "invalid_config",
                f"the cache lifetime of {_name(model)} is unknown: set context.cache_ttl_ms, or "
                "pass cache_ttl_ms to its factory",
            )
        ttl = cache["ttl_ms"]
        if agreed is None:
            agreed = (model, ttl)
        elif agreed[1] != ttl:
            raise ConfigError(
                "invalid_config",
                f"{_name(agreed[0])} caches for {_span(agreed[1])} but fallback {_name(model)} "
                f"for {_span(ttl)}: set context.cache_ttl_ms",
            )
    return None if agreed is None else agreed[1]


def _name(model: Model) -> str:
    ref = model.info.model
    return f"{ref.provider}/{ref.name}"


def _span(ms: int) -> str:
    if ms % 3_600_000 == 0:
        return f"{ms // 3_600_000}h"
    if ms % 60_000 == 0:
        return f"{ms // 60_000}m"
    return f"{ms} ms"
